anndata/tests/test_awkward.py

```diff
@@ -0,0 +1,273 @@
+
+
+@pytest.mark.parametrize("key", ["obsm", "varm", "uns"])
+def test_copy(key):
+    """Check that modifying a copy does not modify the original"""
+    adata = gen_adata((3, 3), varm_types=(), obsm_types=(), layers_types=())
+    getattr(adata, key)["awk"] = ak.Array([{"a": [1], "b": [2], "c": [3]}] * 3)
+    adata_copy = adata.copy()
+    getattr(adata_copy, key)["awk"]["c"] = np.full((3, 1), 4)
+    getattr(adata_copy, key)["awk"]["d"] = np.full((3, 1), 5)
+
+    # values in copy were correctly set
+    npt.assert_equal(getattr(adata_copy, key)["awk"]["c"], np.full((3, 1), 4))
+    npt.assert_equal(getattr(adata_copy, key)["awk"]["d"], np.full((3, 1), 5))
+
+    # values in original were not updated
+    npt.assert_equal(getattr(adata, key)["awk"]["c"], np.full((3, 1), 3))
+    with pytest.raises(IndexError):
+        getattr(adata, key)["awk"]["d"]
+
+
+@pytest.mark.parametrize("key", ["obsm", "varm"])
+def test_view(key):
+    """Check that modifying a view does not modify the original"""
+    adata = gen_adata((3, 3), varm_types=(), obsm_types=(), layers_types=())
+    getattr(adata, key)["awk"] = ak.Array([{"a": [1], "b": [2], "c": [3]}] * 3)
+    adata_view = adata[:2, :2]
+
+    with pytest.warns(ImplicitModificationWarning, match="initializing view as actual"):
+        getattr(adata_view, key)["awk"]["c"] = np.full((2, 1), 4)
+        getattr(adata_view, key)["awk"]["d"] = np.full((2, 1), 5)
+
+    # values in view were correctly set
+    npt.assert_equal(getattr(adata_view, key)["awk"]["c"], np.full((2, 1), 4))
+    npt.assert_equal(getattr(adata_view, key)["awk"]["d"], np.full((2, 1), 5))
+
+    # values in original were not updated
+    npt.assert_equal(getattr(adata, key)["awk"]["c"], np.full((3, 1), 3))
+    with pytest.raises(IndexError):
+        getattr(adata, key)["awk"]["d"]
+
+
+@pytest.mark.parametrize(
+    "array",
+    [
+        # numpy array
+        ak.Array(np.arange(2 * 3 * 4 * 5).reshape((2, 3, 4, 5))),
+        # record
+        ak.Array([{"a": 1, "b": 2}, {"a": 1, "b": 3}]),
+        # ListType, variable length
+        ak.Array([[1], [2, 3], [4, 5, 6]]),
+        # RegularType + nested ListType
+        ak.to_regular(ak.Array([[[1, 2], [3]], [[2], [3, 4, 5]]]), 1),
+        # nested record
+        ak.to_regular(ak.Array([[{"a": 0}, {"b": 1}], [{"c": 2}, {"d": 3}]]), 1),
+        # mixed types (variable length)
+        ak.Array([[1, 2], ["a"]]),
+        # zero-size edge cases
+        ak.Array(np.ones((0, 7))),
+        ak.Array(np.ones((7, 0))),
+        # UnionType of two regular types with different dimensions
+        ak.concatenate([ak.Array(np.ones((2, 2))), ak.Array(np.ones((2, 3)))]),
+        # UnionType of two regular types with same dimension
+        ak.concatenate(
+            [
+                ak.Array(np.ones((2, 2))),
+                ak.Array(np.array([["a", "a"], ["a", "a"]])),
+            ]
+        ),
+    ],
+)
+def test_awkward_io(tmp_path, array):
```

Comment:
Do you want to support categorical values in awkward arrays? If so, that would be a good test case.

Reply:
You mean something with `ak.to_categorical`?

Reply:
Yeah, if that's something you'd like to use for your `airr` data.

I would imagine that some of the fields could be encoded as a categorical, like locus names.

Reply:
That would replace an array of strings (at some level of the hierarchy) with an `IndexedArray` of strings, like this:

```xml
<IndexedArray len="1000000">
    <parameter name="__array__">categorical</parameter>
    <Index len=1000000">3, 1, 0, 2, 1, 2, 0, 1, ..., 2, 1, 1</Index>
    ... array of 4 unique strings ...
</IndexedArray>
```

(pseudocode above). So if you have a large set of strings for which only a few are unique, this can reduce memory and speed up comparisons. If I remember right, `==` is overloaded to do an integer comparison, rather than a string comparison (and if I'm remembering wrong and that hasn't been implemented yet, it can be).

The reason for the parameter `__array__` = `"categorical"` is so that this structure is converted into an Arrow/Parquet [dictionary encoding](https://arrow.apache.org/blog/2019/09/05/faster-strings-cpp-parquet/) on write. Without the parameter declaring that the strings are actually unique, the `to_arrow`/`to_parquet` conversion conservatively assumes that they're not unique (and Arrow or Parquet _requires_ a dictionary to contain unique values).

Reply:
@jpivarski, what would be the `v2` equivalent of `to_categorical`?

Reply:
I just checked on this, `to_categorical` and `from_categorical` [have been implemented](https://github.com/scikit-hep/awkward/blob/1bcfd70629aec4c1ef26bbf929d9183c28dbe491/src/awkward/_v2/behaviors/categorical.py#L162-L327), but they haven't been exposed in the `ak._v2` module yet. (?) It should be accessible as `ak._v2.behaviors.categorical.to_categorical` until we get that fixed.

Reply:
@agoose77 had the better idea of moving these functions into the standard place: `ak._v2.operations.*categori*` (https://github.com/scikit-hep/awkward/pull/1674). After the next release, that's where they'll be, as well as `ak.*categori*`.